src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -48,6 +51,288 @@ let GetInitialOptimizationEnv (tcImports: TcImports, tcGlobals: TcGlobals) =
     let optEnv = List.fold (AddExternalCcuToOptimizationEnv tcGlobals) optEnv ccuinfos
     optEnv
 
+[<RequireQualifiedAccess>]
+module private ParallelOptimization =
+    open Optimizer
+    type OptimizeDuringCodeGen = bool -> Expr -> Expr
+
+    type OptimizeRes =
+        (IncrementalOptimizationEnv * CheckedImplFile * ImplFileOptimizationInfo * SignatureHidingInfo) * OptimizeDuringCodeGen
+
+    type PhaseInputs = IncrementalOptimizationEnv * SignatureHidingInfo * CheckedImplFile
+    type Phase1Inputs = PhaseInputs
+    type Phase1Res = OptimizeRes
+    type Phase1Fun = Phase1Inputs -> Phase1Res
+    type Phase2Inputs = IncrementalOptimizationEnv * SignatureHidingInfo * ImplFileOptimizationInfo * CheckedImplFile
+    type Phase2Res = IncrementalOptimizationEnv * CheckedImplFile
+    type Phase2Fun = Phase2Inputs -> Phase2Res
+    type Phase3Inputs = PhaseInputs
+    type Phase3Res = IncrementalOptimizationEnv * CheckedImplFile
+    type Phase3Fun = Phase3Inputs -> Phase3Res
+
+    type FileResultsComplete =
+        {
+            Phase1: Phase1Res
+            Phase2: Phase2Res
+            Phase3: Phase3Res
+        }
+
+    type FilePhaseFuncs = Phase1Fun * Phase2Fun * Phase3Fun
+
+    [<RequireQualifiedAccess>]
+    type OptimizationPhase =
+        | Phase1
```

Comment:
I would suggest that you make a comment on what these phases are going to do.

Reply:
I have replaced this with a dynamic list of named phases - names are now next to where the phase functions are defined in `ApplyAllOptimizations`.